eth/watchers/eventdecoder.go

```diff
@@ -0,0 +1,73 @@
+package watchers
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// EventDecoder decodes logs into events for known contracts
+type EventDecoder struct {
+	addr             ethcommon.Address
+	contract         *bind.BoundContract
+	topicToEventName map[ethcommon.Hash]string
+}
+
+// FindEventName returns the event name for a log. An error will be returned if the log is not emitted
+// from a known contract or if it does not map to a known event
+func (e *EventDecoder) FindEventName(log types.Log) (string, error) {
+	switch log.Address {
+	case e.addr:
+		return e.findEventName(log)
+	default:
+		return "", errors.New("log not from known contract")
+	}
+}
+
+// Decode decodes a log into an event struct. An error will be returned if the log is not emitted
+// from a known contract or if it does not map to a known event
+func (e *EventDecoder) Decode(eventName string, log types.Log, decodedLog interface{}) error {
+	switch log.Address {
+	case e.addr:
+		return e.decode(eventName, log, decodedLog)
+	default:
+		return errors.New("log not from known contract")
+	}
+}
+
+func NewEventDecoder(addr ethcommon.Address, abiJSON string) (*EventDecoder, error) {
+	abi, err := abi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return nil, err
+	}
+
+	topicToEventName := make(map[ethcommon.Hash]string)
+	for _, event := range abi.Events {
+		topicToEventName[event.Id()] = event.Name
+	}
+
+	return &EventDecoder{
+		addr: addr,
+		// Create BoundContract without a backend because we just need to access
+		// log unpacking without contract interaction
+		contract:         bind.NewBoundContract(addr, abi, nil, nil, nil),
+		topicToEventName: topicToEventName,
+	}, nil
+}
+
+func (e *EventDecoder) decode(eventName string, log types.Log, decodedLog interface{}) error {
+	return e.contract.UnpackLog(decodedLog, eventName, log)
+}
+
+func (e *EventDecoder) findEventName(log types.Log) (string, error) {
+	eventName, ok := e.topicToEventName[log.Topics[0]]
+	if !ok {
+		return "", errors.New("unknown event for BondingManager")
```

Comment:
We should update this error message since `EventDecoder` is no longer specific to the BondingManager contract. Instead of providing an explicit name, since `EventDecoder` already has access to the contract address, we could change the error to something like this:

```
return "", fmt.Errorf("unknown event for %v", e.addr.Hex())
```

Reply:
fixed in ca5b5af